no-var --fix sometimes leads to ReferenceError

**Tell us about your environment**

Short summary of the issue:

1. `var x = x = 1`
1. `eslint --fix`
1. `let x = x = 1`
1. `ReferenceError: x is not defined`

* **ESLint Version:** 3.13.1
* **Node Version:** 6.9.1
* **npm Version:** 4.1.1

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<!-- Paste your configuration below: -->
```json
{
  "parserOptions": {
    "ecmaVersion": 8,
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true,
      "jsx": true
    },
    "sourceType": "module"
  },

  "env": {
    "es6": true,
    "node": true,
    "mocha": true
  },

  "plugins": [],

  "globals": {
    "document": false,
    "navigator": false,
    "window": false,
    "expect": true,
    "test": true,
    "sinon": true
  },

  "rules": {
    "accessor-pairs": 2,
    "arrow-spacing": [2, {"before": true, "after": true }],
    "block-spacing": [2, "always"],
    "brace-style": [2, "1tbs", {"allowSingleLine": true }],
    "camelcase": [2, {"properties": "never"}],
    "comma-dangle": [2, "never"],
    "comma-spacing": [2, {"before": false, "after": true }],
    "comma-style": [2, "last"],
    "computed-property-spacing": [2, "never"],
    "constructor-super": 2,
    "curly": [2, "multi-line"],
    "dot-location": [2, "property"],
    "eol-last": [2, "always"],
    "eqeqeq": [2, "allow-null"],
    "func-call-spacing": [2, "never"],
    "handle-callback-err": [2, "^(err|error)$" ],
    "indent": [2, 2, {
      "SwitchCase": 1,
      "VariableDeclarator": 1,
      "outerIIFEBody": 1,
      "FunctionDeclaration": {
        "parameters": 1,
        "body": 1
      },
      "FunctionExpression": {
        "parameters": 1,
        "body": 1
      }
    }],
    "key-spacing": [2, {"beforeColon": false, "afterColon": true }],
    "keyword-spacing": [2, {"before": true, "after": true }],
    "linebreak-style": [2, "unix"],
    "max-len": ["error", {
      "code": 100,
      "ignoreRegExpLiterals": true,
      "ignorePattern": "^\\s*const\\s.+=\\s*require\\s*\\(/"
    }],
    "new-cap": [2, {"newIsCap": true, "capIsNew": false }],
    "new-parens": 2,
    "newline-per-chained-call": [2, {"ignoreChainWithDepth": 4 }],
    "no-array-constructor": 2,
    "no-caller": 2,
    "no-class-assign": 2,
    "no-cond-assign": 2,
    "no-const-assign": 2,
    "no-constant-condition": [2, {"checkLoops": false }],
    "no-control-regex": 2,
    "no-debugger": 2,
    "no-delete-var": 2,
    "no-dupe-args": 2,
    "no-dupe-class-members": 2,
    "no-dupe-keys": 2,
    "no-duplicate-case": 2,
    "no-duplicate-imports": 2,
    "no-empty-character-class": 2,
    "no-empty-pattern": 2,
    "no-eval": 2,
    "no-ex-assign": 2,
    "no-extend-native": 2,
    "no-extra-bind": 2,
    "no-extra-boolean-cast": 2,
    "no-extra-parens": [2, "functions"],
    "no-fallthrough": 2,
    "no-floating-decimal": 2,
    "no-func-assign": 2,
    "no-global-assign": 2,
    "no-implied-eval": 2,
    "no-inner-declarations": [2, "functions"],
    "no-invalid-regexp": 2,
    "no-irregular-whitespace": 2,
    "no-iterator": 2,
    "no-label-var": 2,
    "no-labels": [2, {"allowLoop": false, "allowSwitch": false }],
    "no-lone-blocks": 2,
    "no-lonely-if": 2,
    "no-mixed-operators": [2, {
      "groups": [
        ["+", "-", "*", "/", "%", "**"],
        ["&", "|", "^", "~", "<<", ">>", ">>>"],
        ["==", "!=", "===", "!==", ">", ">=", "<", "<="],
        ["&&", "||"],
        ["in", "instanceof"]
      ],
      "allowSamePrecedence": false
    }],
    "no-mixed-spaces-and-tabs": 2,
    "no-multi-spaces": 2,
    "no-multi-str": 2,
    "no-multiple-empty-lines": [2, {"max": 2 }],
    "no-native-reassign": 2,
    "no-negated-in-lhs": 2,
    "no-nested-ternary": 2,
    "no-new": 2,
    "no-new-func": 2,
    "no-new-object": 2,
    "no-new-require": 2,
    "no-new-symbol": 2,
    "no-new-wrappers": 2,
    "no-obj-calls": 2,
    "no-octal": 2,
    "no-octal-escape": 2,
    "no-path-concat": 2,
    "no-proto": 2,
    "no-redeclare": 2,
    "no-regex-spaces": 2,
    "no-return-assign": [2, "except-parens"],
    "no-self-assign": 2,
    "no-self-compare": 2,
    "no-sequences": 2,
    "no-shadow-restricted-names": 2,
    "no-sparse-arrays": 2,
    "no-tabs": 2,
    "no-template-curly-in-string": 2,
    "no-this-before-super": 2,
    "no-throw-literal": 2,
    "no-trailing-spaces": 2,
    "no-undef": 2,
    "no-undef-init": 2,
    "no-unexpected-multiline": 2,
    "no-unmodified-loop-condition": 2,
    "no-unneeded-ternary": [2, {"defaultAssignment": false }],
    "no-unreachable": 2,
    "no-unsafe-finally": 2,
    "no-unsafe-negation": 2,
    "no-unused-vars": [2, {"vars": "all", "args": "none"}],
    "no-useless-call": 2,
    "no-useless-computed-key": 2,
    "no-useless-constructor": 2,
    "no-useless-escape": 2,
    "no-useless-rename": 2,
    "no-var": 2,
    "no-whitespace-before-property": 2,
    "no-with": 2,
    "object-curly-spacing": [2, "never"],
    "object-property-newline": [2, {"allowMultiplePropertiesPerLine": true }],
    "one-var": [2, {"initialized": "never"}],
    "operator-linebreak": [2, "after", {"overrides": {"?": "before", ":": "before"}}],
    "padded-blocks": [0, "never"],
    "prefer-template": 2,
    "prefer-const": ["error", {"destructuring": "any", "ignoreReadBeforeAssign": true}],
    "quotes": [2, "single", {"avoidEscape": true, "allowTemplateLiterals": true }],
    "rest-spread-spacing": [2, "never"],
    "semi": [2, "never"],
    "semi-spacing": [2, {"before": false, "after": true }],
    "space-before-blocks": [2, "always"],
    "space-before-function-paren": [2, "always"],
    "space-in-parens": [2, "never"],
    "space-infix-ops": 2,
    "space-unary-ops": [2, {"words": true, "nonwords": false }],
    "spaced-comment": [2, "always", {"line": {"markers": ["*package", "!", ","]}, "block": {"balanced": true, "markers": ["*package", "!", ","], "exceptions": ["*"]}}],
    "template-curly-spacing": [2, "never"],
    "unicode-bom": [2, "never"],
    "use-isnan": 2,
    "valid-typeof": 2,
    "wrap-iife": [2, "any", {"functionPrototypeMethods": true }],
    "yield-star-spacing": [2, "both"],
    "yoda": [2, "never"]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
$ echo 'var x = x = 1' > test.js
$ node test.js
$ eslint --fix test.js
$ cat test.js
let x = x = 1
```

**What did you expect to happen?**

I expected it to report the error below, which is what I get with the following minimal `.eslintrc`:

```json
{
  "rules": {
    "no-var": 2
  }
}
```

```js
/test.js
  1:5  error  Parsing error: Unexpected token x

✖ 1 problem (1 error, 0 warnings)
```

**What actually happened? Please include the actual, raw output from ESLint.**

```js
$ eslint --fix test.js
/test.js
  1:5  error  'x' is assigned a value but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

$ node test.js
/test.js:1
(function (exports, require, module, __filename, __dirname) { let x = x = 1
                                                                        ^

ReferenceError: x is not defined
    at Object.<anonymous> (/test.js:1:73)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

--- a/lib/rules/no-var.js
+++ b/lib/rules/no-var.js
@@ -128,6 +128,43 @@ function isUsedFromOutsideOf(scopeNode) {
     };
 }
 
+/**
+ * Creates the predicate function which checks whether a variable has their references in TDZ.
+ *
+ * The predicate function would return `true`:
+ *
+ * - if a reference is before the declarator. E.g. (var a = b, b = 1;)(var {a = b, b} = {};)
+ * - if a reference is in the expression of their default value.  E.g. (var {a = a} = {};)
+ * - if a reference is in the expression of their initializer.  E.g. (var a = a;)
+ *
+ * @param {ASTNode} node - The initializer node of VariableDeclarator.
+ * @returns {Function} The predicate function.
+ * @private
+ */
+function hasReferenceInTDZ(node) {
+    const initStart = node.range[0];
+    const initEnd = node.range[1];
+
+    return variable => {
+        const id = variable.defs[0].name;
+        const idStart = id.range[0];
+        const defaultValue = (id.parent.type === "AssignmentPattern" ? id.parent.right : null);
+        const defaultStart = defaultValue && defaultValue.range[0];
+        const defaultEnd = defaultValue && defaultValue.range[1];
+
+        return variable.references.some(reference => {
+            const start = reference.identifier.range[0];
+            const end = reference.identifier.range[1];
+
+            return !reference.init && (
+                start < idStart ||
+                (defaultValue !== null && start >= defaultStart && end <= defaultEnd) ||
+                (start >= initStart && end <= initEnd)
+            );
+        });
+    };
+}
+
 //------------------------------------------------------------------------------
 // Rule Definition
 //------------------------------------------------------------------------------
@@ -148,6 +185,21 @@ module.exports = {
         const sourceCode = context.getSourceCode();
 
         /**
+         * Checks whether the variables which are defined by the given declarator node have their references in TDZ.
+         *
+         * @param {ASTNode} declarator - The VariableDeclarator node to check.
+         * @returns {boolean} `true` if one of the variables which are defined by the given declarator node have their references in TDZ.
+         */
+        function hasSelfReferenceInTDZ(declarator) {
+            if (!declarator.init) {
+                return false;
+            }
+            const variables = context.getDeclaredVariables(declarator);
+
+            return variables.some(hasReferenceInTDZ(declarator.init));
+        }
+
+        /**
          * Checks whether it can fix a given variable declaration or not.
          * It cannot fix if the following cases:
          *
@@ -156,6 +208,7 @@ module.exports = {
          * - A variable is used from outside the scope.
          * - A variable is used from a closure within a loop.
          * - A variable might be used before it is assigned within a loop.
+         * - A variable might be used in TDZ.
          *
          * ## A variable is declared on a SwitchCase node.
          *
@@ -201,8 +254,10 @@ module.exports = {
             const scopeNode = getScopeNode(node);
 
             if (node.parent.type === "SwitchCase" ||
-                    variables.some(isRedeclared) ||
-                    variables.some(isUsedFromOutsideOf(scopeNode))) {
+                node.declarations.some(hasSelfReferenceInTDZ) ||
+                variables.some(isRedeclared) ||
+                variables.some(isUsedFromOutsideOf(scopeNode))
+            ) {
                 return false;
             }
 
@@ -241,7 +296,7 @@ module.exports = {
         }
 
         return {
-            VariableDeclaration(node) {
+            "VariableDeclaration:exit"(node) {
                 if (node.kind === "var") {
                     report(node);
                 }

--- a/tests/lib/rules/no-var.js
+++ b/tests/lib/rules/no-var.js
@@ -154,6 +154,67 @@ ruleTester.run("no-var", rule, {
             errors: [
                 "Unexpected var, use let or const instead."
             ]
+        },
+
+        // https://github.com/eslint/eslint/issues/7950
+        {
+            code: "var a = a",
+            output: "var a = a",
+            errors: [
+                "Unexpected var, use let or const instead."
+            ]
+        },
+        {
+            code: "var {a = a} = {}",
+            output: "var {a = a} = {}",
+            parserOptions: { ecmaVersion: 2015 },
+            errors: [
+                "Unexpected var, use let or const instead."
+            ]
+        },
+        {
+            code: "var {a = b, b} = {}",
+            output: "var {a = b, b} = {}",
+            parserOptions: { ecmaVersion: 2015 },
+            errors: [
+                "Unexpected var, use let or const instead."
+            ]
+        },
+        {
+            code: "var {a, b = a} = {}",
+            output: "let {a, b = a} = {}",
+            parserOptions: { ecmaVersion: 2015 },
+            errors: [
+                "Unexpected var, use let or const instead."
+            ]
+        },
+        {
+            code: "var a = b, b = 1",
+            output: "var a = b, b = 1",
+            parserOptions: { ecmaVersion: 2015 },
+            errors: [
+                "Unexpected var, use let or const instead."
+            ]
+        },
+        {
+            code: "var a = b; var b = 1",
+            output: "let a = b; var b = 1",
+            parserOptions: { ecmaVersion: 2015 },
+            errors: [
+                "Unexpected var, use let or const instead.",
+                "Unexpected var, use let or const instead."
+            ]
+        },
+
+        // This case is not in TDZ, but it's very hard to distinguish the reference is in TDZ or not.
+        // So this rule does not fix it for safe.
+        {
+            code: "function foo() { a } var a = 1; foo()",
+            output: "function foo() { a } var a = 1; foo()",
+            parserOptions: { ecmaVersion: 2015 },
+            errors: [
+                "Unexpected var, use let or const instead."
+            ]
         }
     ]
 });

